### DISCUSSION TITLE
Compile-time match expression

### DISCUSSION THREAD

**Discussion by refrogerator** [vlang/v#11825] [Ideas]
We currently only have compile-time if statetemts to check for a type in situations like these:
```v
       $if field.typ is string {
            // $(string_expr) produces an identifier
            result.$(field.name) = get_string(data, field.name)
        } $else $if field.typ is int {
            result.$(field.name) = get_int(data, field.name)
        }
```
A compile-time match expression is way better suited for this and more readable imho.

**Comment by ntrel** [comment]
I think compile-time `match` might be quite a burden on the compiler, when it does the same thing as `$if`. Consider that the condition for `$match` can be a type, or it can be a compile-time value - that's two fundamentally different forms. Then `match` has various syntax for patterns, it all gets quite complex.

BTW if we implement #8543 and generalise `$(string_expr)` -> identifier, your example could be 3 lines, even handling more types:
```v
$if field.Type in string | int | rune | bool {
    result.$(field.name) = $('get_' + field.Type.name)(data, field.name)
}
```

**Comment by medvednikov** [comment]
I agree with @ntrel 

I feel like having just compile time $ifs will make the language simpler.

**Comment by JalonSolov** [comment]
Simpler, but a lot more verbose if you have a complex if/else if/else if/else if/else if/...

Just think of how "messy" a large if/else block looks compared to a `switch/case` block.

**Comment by Cons-Cat** [comment]
I personally feel like it's "simple" for any runtime expression to be easily factored into a comptime expression assuming that the values it operates on are all compile-time known.

If `$if`, `$else`, `$for`, and `$match` are all valid, then the rule I need to memorize is "expressions can be made compile-time by prefixing $".

If `$match` is not valid, then I need to memorize that match is an exception to the above rule, so the rules of V are more complex.
